downloads: minimize syscalls to improve performance

pohly opened the pull request:
This fixes https://github.com/clearlinux/swupd-client/issues/41

Beware that I have only tested a slightly different patch against
2.87. Porting it to the current master code required some manual
conflict resolution and I haven't had the time to try the result.

barbieri commented on src/curl.c:
```diff
@@ -157,39 +157,58 @@ static size_t swupd_download_version_to_memory(void *ptr, size_t size, size_t nm
+        for (remaining = size*nmemb; remaining; remaining -= written) {
+		written = write(fd, ptr, size*nmemb);
+		if (written < 0) {
+			LOG_ERROR(file, "write error", class_file_io,
```
write() can fail on `EINTR`, which is not a fatal error and you could try again

pohly replied:
True, need to add that.

barbieri commented on include/swupd.h:
```diff
@@ -84,6 +84,7 @@ struct file {
 	int last_change;
 	struct update_stat stat;
 
+	unsigned int fd_valid : 1;
```
no need for this, once you `close(fd)`, make the `fd = -1` and use that as `fd_valid`

pohly replied:
The problem is that there is no explicit initializer function for the struct where one could set -1 reliably in all places; one can only rely on members being cleared (calloc). That's why "fd_valid" != 0 works, but "fd != -1" does not.

barbieri replied:
well.. then use zero. While zero is a valid fd, in this scenario it will never happen for the file you're writing. At least you have the one that is providing you with the data (socket or local file), also the swupd-client is not closing stdout/stderr, which are used to log and communicate with the user.

phmccarty replied:
@pohly I am open to adding file initializer helper functions, for both a struct file on the stack, and dynamically allocated.

phmccarty replied:
@pohly @rojkov Any thoughts on this approach? I really think fd initialized to -1 would be better than using the separate fd_valid field.

phmccarty commented on src/curl.c:
```diff
@@ -157,39 +157,58 @@ static size_t swupd_download_version_to_memory(void *ptr, size_t size, size_t nm
+	size_t written, remaining;
 
 	outfile = file->staging;
+        if (file->fd_valid) {
```
The swupd codebase uses tabs+spaces for indentation.

phmccarty commented on src/curl.c:
```diff
@@ -157,39 +157,58 @@ static size_t swupd_download_version_to_memory(void *ptr, size_t size, size_t nm
+		fd = open(outfile, O_CREAT | O_RDWR | O_CLOEXEC | O_APPEND, 00600);
+		if (fd < 0) {
+			LOG_ERROR(file, "Cannot open file for write", class_file_io,
+				  "\\*outfile=\"%s\",strerror=\"%s\"*\\", outfile, strerror(errno));
```
LOG_* macros are not longer used, so this can just be a printf().

phmccarty commented on src/curl.c:
```diff
@@ -157,39 +157,58 @@ static size_t swupd_download_version_to_memory(void *ptr, size_t size, size_t nm
 size_t swupd_download_file(void *ptr, size_t size, size_t nmemb, void *userdata)
 {
-	struct file *file = (struct file *)userdata;
+	struct file *file = (struct file*)userdata;
```
The space before * is the style enforced.

phmccarty commented on src/download.c:
```diff
@@ -164,6 +164,7 @@ static void free_curl_list_data(void *data)
 {
 	struct file *file = (struct file *)data;
 	CURL *curl = file->curl;
+	swupd_download_file_complete(CURLE_OK, file);
```
The codebase is not consistent here, but I would like to see an explicit (void) cast to indicate an unused return value.

phmccarty commented on src/curl.c:
```diff
@@ -227,7 +246,6 @@ int swupd_curl_get_file(const char *url, char *filename, struct file *file,
 			}
 		}
 		local->staging = filename;
-
```
extra line removed?

phmccarty commented on src/curl.c:
```diff
@@ -157,39 +157,58 @@ static size_t swupd_download_version_to_memory(void *ptr, size_t size, size_t nm
+        }
+
+        /* handle short writes with repeated write() calls */
+        for (remaining = size*nmemb; remaining; remaining -= written) {
```
For style, we're adding spaces before and after all arithmetic operators.

barbieri replied:
In soletta we provide a check-style script that calls uncrustify and if diff is non-0 we complain. Recommend doing so and using a semaphore-ci to ensure that on PR.

barbieri replied:
https://github.com/solettaproject/soletta/blob/master/tools/check-style.py

phmccarty commented on include/swupd.h:
```diff
@@ -104,6 +105,7 @@ struct file {
 
 	char *staging; /* output name used during download & staging */
 	CURL *curl;    /* curl handle if downloading */
+	int fd;        /* file written into during downloading, unset when fd_valid is false */
```
I would prefer to use fd == -1 to indicate an invalid file descriptor, with the requisite addition of struct file initializers.

pohly replied:
On Thu, 2016-04-21 at 14:19 -0700, Patrick McCarty wrote:
> In include/swupd.h:
> 
> > @@ -104,6 +105,7 @@ struct file {
> >  
> >  	char *staging; /* output name used during download & staging */
> >  	CURL *curl;    /* curl handle if downloading */
> > +	int fd;        /* file written into during downloading, unset when fd_valid is false */
> 
> I would prefer to use fd == -1 to indicate an invalid file descriptor,
> with the requisite addition of struct file initializers.

That would indeed be the cleaner solution, but introducing such an
initializer in all the right places IMHO goes beyond the scope of this
patch and should better be done by someone who knows the code.

So in other words, it would be good if someone could do that, but please
don't expect me or Dmitry to do it ;-}

phmccarty replied:
That's fair. I will add it to my TODO list then.

bryteise commented:
How about an alternative that uses the default curl handler for CURLOPT_WRITEFUNCTION and open the FILE * prior to initiating the download?

phmccarty commented:
@bryteise That sounds reasonable, I think. So we only need to set CURLOPT_WRITEDATA to the FILE * we open for file->staging, because the default callback is fwrite()...

pohly commented:
+1 for using the default curl write handler. I didn't do that because I wanted to keep the existing semantic (= file only opened when transfer at least has started), but if that's not a problem, then it's certainly easier to open in advance.

Can you guys fix it that way? I'm a bit swamped right now.

bryteise commented:
So @phmccarty attempted the default curl write handler and because of how we queue downloads we end up with too many open file descriptors so it seems like @pohly has the best approach.

phmccarty commented:
@bryteise Right. I forgot to post the latest status here. The current async download code flow does not work well with the default write handler, so I would need to refactor so make it work.

pohly commented:
@rojkov was so kind to address the review feedback (https://github.com/pohly/swupd-client/pull/1).

As this is now seen as the right approach, I merged those fixes, squashed, rebased and force-pushed them in this PR. Please consider merging. No manual testing by myself, but @rojkov ran tests.

phmccarty commented on src/curl.c:
```diff
@@ -169,35 +169,57 @@ size_t swupd_download_file(void *ptr, size_t size, size_t nmemb, void *userdata)
+		fd = open(outfile, O_CREAT | O_RDWR | O_CLOEXEC | O_APPEND, 00600);
+		if (fd < 0) {
+			fprintf(stderr, "Cannot open file for write \\*outfile=\"%s\",strerror=\"%s\"*\\\n",
+				outfile, strerror(errno));
```
When testing this patch for a large update of Clear Linux, I am running into this error condition repeatedly. The specific open(2) error is EMFILE (Too many open files).

I think there is a bug in the hysteresis logic where too many files are being added to the download queue and not being processed fast enough, thus reaching the max open fd limit when the files are open for the entirety of the download. I will debug this issue first.

pohly commented:
@phmccarty when testing this change in May, you ran into a "bug in the hysteresis logic" which you wanted to debug. Has that been fixed? Can this PR be merged now?

phmccarty commented:
@pohly The hysteresis logic fix is merged (see #133).

I would rather use the default curl write handler (fwrite(3)) when downloading to files, as discussed in previous comments on this PR. But I haven't yet investigated the proper fix to use the default handler everywhere in the code. Some refactoring of the download code may be needed as well.

pohly commented:
@phmccarty @tmarcu @bryteise: I've changed the code such that it uses the default curl write handler. I've intentionally not squashed or rebased, because the old commit was working and preserving it IMHO is useful.

Please review and merge.

phmccarty commented:
@pohly I agree that keeping the original commit is helpful.

+1 from me. Thanks!
